pallets/msa/src/tests.rs

```diff
@@ -645,6 +645,114 @@ pub fn add_provider_to_msa_is_success() {
+
+		// Create delegator account and get its MSA ID (u64)
+		assert_ok!(Msa::create(RuntimeOrigin::signed(delegator_account.into())));
+		let delegator_msa =
+			Msa::ensure_valid_msa_key(&AccountId32::new(delegator_account.0)).unwrap();
+
+		// Register provider
+		assert_ok!(Msa::create_provider(
+			RuntimeOrigin::signed(provider_account.into()),
+			Vec::from("Foo")
+		));
+
+		let (delegator_signature, add_provider_payload) =
+			create_and_sign_add_provider_payload(delegator_pair.clone(), provider_msa);
+
+		set_schema_count::<Test>(10);
+
+		assert_ok!(Msa::grant_delegation(
+			RuntimeOrigin::signed(provider_account.into()),
+			delegator_account.into(),
+			delegator_signature,
+			add_provider_payload
+		));
+
+		let provider = ProviderId(provider_msa);
+		let delegator = DelegatorId(delegator_msa);
+
+		assert_eq!(
+			Msa::get_delegation(delegator, provider),
+			Some(Delegation { revoked_at: 0, schema_permissions: Default::default() })
+		);
+
+		System::assert_last_event(
+			Event::DelegationGranted {
+				delegator_id: delegator_msa.into(),
+				provider_id: provider_msa.into(),
+			}
+			.into(),
+		);
+
+		// Grant delegation w/schemas 1, 2, 3, and 4.
+		let (delegator_signature, add_provider_payload) =
+			create_and_sign_add_provider_payload_with_schemas(
+				delegator_pair.clone(),
+				provider_msa,
+				Some(vec![1, 2, 3, 4]),
+			);
+
+		assert_ok!(Msa::grant_delegation(
+			RuntimeOrigin::signed(provider_account.into()),
+			delegator_account.into(),
+			delegator_signature,
+			add_provider_payload
+		));
+
+		let mut sp = BoundedBTreeMap::<SchemaId, u64, MaxSchemaGrantsPerDelegation>::new();
+		assert_ok!(sp.try_insert(1u16, 0u64));
+		assert_ok!(sp.try_insert(2u16, 0u64));
+		assert_ok!(sp.try_insert(3u16, 0u64));
+		assert_ok!(sp.try_insert(4u16, 0u64));
+
+		let expected = Delegation { revoked_at: 0, schema_permissions: sp };
+
+		assert_eq!(Msa::get_delegation(delegator, provider), Some(expected));
+
+		// Grant delegation w/schemas 3, 4, 5, and 6.
+		// This should add 5 and 6 and revoke 1 and 2.
+		let (delegator_signature, add_provider_payload) =
+			create_and_sign_add_provider_payload_with_schemas(
+				delegator_pair.clone(),
+				provider_msa,
+				Some(vec![3, 4, 5, 6]),
+			);
+
+		assert_ok!(Msa::grant_delegation(
+			RuntimeOrigin::signed(provider_account.into()),
+			delegator_account.into(),
+			delegator_signature,
+			add_provider_payload
+		));
+
+		let mut sp = BoundedBTreeMap::<SchemaId, u64, MaxSchemaGrantsPerDelegation>::new();
+		assert_ok!(sp.try_insert(1u16, 1u64)); // schema id 1 revoked at block 1
+		assert_ok!(sp.try_insert(2u16, 1u64)); // schema id 2 revoked at block 1
+		assert_ok!(sp.try_insert(3u16, 0u64));
+		assert_ok!(sp.try_insert(4u16, 0u64));
+		assert_ok!(sp.try_insert(5u16, 0u64));
+		assert_ok!(sp.try_insert(6u16, 0u64));
+
+		let expected = Delegation { revoked_at: 0, schema_permissions: sp };
+
+		assert_eq!(Msa::get_delegation(delegator, provider), Some(expected));
```

Comment:
I just added the following to the end of this test, which makes it fail.  So I'm not sure if this is the correct expectation. I would think that when sending a new set of permissions, schemas that are already revoked should _not_ have the expiration block updated. Secondly I expected to be able to re-permit/renew a previously expired schema:

```

		System::set_block_number(5);
		let (delegator_signature, add_provider_payload) =
			create_and_sign_add_provider_payload_with_schemas(
				delegator_pair.clone(),
				provider_msa,
				Some(vec![1, 3, 6]),
			);

		assert_ok!(Msa::grant_delegation(
			RuntimeOrigin::signed(provider_account.into()),
			delegator_account.into(),
			delegator_signature,
			add_provider_payload
		));

		let mut sp = BoundedBTreeMap::<SchemaId, u64, MaxSchemaGrantsPerDelegation>::new();
		assert_ok!(sp.try_insert(1u16, 0u64)); // schema id 1 should be renewed
		assert_ok!(sp.try_insert(2u16, 1u64)); // schema id 2 should stay revoked at block 1
		assert_ok!(sp.try_insert(3u16, 0u64)); // leave alone
		assert_ok!(sp.try_insert(4u16, 5u64)); // revoke
		assert_ok!(sp.try_insert(5u16, 5u64)); // revoke
		assert_ok!(sp.try_insert(6u16, 0u64)); // leave alone

		let exp = Delegation { revoked_at: 5, schema_permissions: sp };
		assert_eq!(Some(exp), Msa::get_delegation(delegator, provider), );
```

In this portion of the test, schema 2 has its expiration block updated to block 5.  Schema 1 is not re-permitted/renewed.

Reply:
My test was set up for block 0.  I can change it to use the current block and I think it will pass.

Reply:
I think there is a bug but I have to change your test because you reversed the assert from what I used in the prior two calls.  There's also a bug in your test.  The `revoked_at` is for revoking all permissions and is only mutated/changed by `revoke_provider() `and not `grant_delegation()`.

Reply:
If the schema is not in the new grant delegation list but is in the current list, it is revoked at the current block.  It needs to do a further test to not include blocks that were already revoked at a previous block so that they stay revoked at the original block and not the current one.

Reply:
Fixed in next update.